[aarch64] add sbgemm inner product op

snadampal opened the pull request:
# Description
Added sbgemm inner product op to enable PyTorch torch.compile() and bf16 fastmath kernels to work together on aarch64.

Fixes # (github issue)

# Checklist

## General

- [x ] Do all unit and benchdnn tests (`make test` and `make test_benchdnn_*`) pass locally for each commit?
same pass rate as in main branch.
- [x] Have you formatted the code using clang-format?

## Performance improvements
1. torch.compile() with fp32 inner product kernel showed up to 5.8x perf improvement on AWS c7g instance
2. torch.compile() with bf16 inner product kernel showed another 1.25x perf improvement on AWS c7g instance
- [ ] Have you submitted performance data that demonstrates performance improvements?

### New features

- [ ] Have you published an RFC for the new feature?
- [ ] Was the RFC approved?
- [ ] Have you added relevant tests?

### Bug fixes

- [ ] Have you included information on how to reproduce the issue (either in a github issue or in this PR)?
- [ ] Have you added relevant regression tests?

## [RFC](https://github.com/oneapi-src/oneDNN/tree/rfcs) PR

- [ ] Does RFC document follow the [template](https://github.com/oneapi-src/oneDNN/blob/rfcs/rfcs/template.md#onednn-design-document-rfc)?
- [ ] Have you added a link to the rendered document?

snadampal commented:
Hi @jondea , @nSircombe , @cfRod , appreciate if any of you can review this PR? thank you!

jondea commented:
HI @snadampal , thanks for the patch! I'm just looking into this now. Just so I can test it, what benchdnn run lines are you using to test, and in particular which ones do you want to enable with this patch?

snadampal commented:
Hi @jondea , I have tested oneDNN inner product tests in fastmath and normal mode to make sure no regressions. The main addition in this PR is to allow `bfloat16` weights format for primitive creation and reuse the matmul sgemm operator. this is possible with fixed format kernels where, in the fast math mode, it will accept bf16 weights to gemm operators.
I've also tested bert model inference using torch.compile() and oneDNN fastmath model `DNNL_DEFAULT_FPMATH_MODE=BF16`.
Now I'm thinking if I need to add any new test scenarios for oneDNN.

snadampal commented:
Hi @jondea , while I check if new unit tests to be added for the sbgemm ops, can you please review and comment? thank you!

jondea commented:
Sorry for the slow reply. It looks great in principle, I'm glad it's such a simple change to enable this!

My only concern is removing the check that the memory format is any. I thought that oneDNN expects you to change the memory format only if it is set to any? For reference, `acl_utils::reorder_to_weight_format` mutates the memory descriptor.

On the PyTorch side, are you planning to pass any or a specific format?

snadampal commented:
Hi @jondea , thanks for the review. I know ACL expects `format_any` so that it can reorder and pack them to the custom format that ACL gemm understands. However, with the torch.compile() and weights pre-packing done before the primitive creation, the weights are already reordered by ACL (even before the primitive is created). but your concern is valid, because the same primitive is used for compile as well as non-compile path.  so, how about we check for either`format_any` or `ACL custom format (blocked:AB8a4b::f0) `

I used the above blocked format because this is what I see during the ACL reorders for fastmath bf16 weights:
`onednn_verbose,primitive,exec,cpu,reorder,jit:uni,undef,src_f32::blocked:ab::f0 dst_bf16::blocked:AB8a4b::f0,attr-scratchpad:user attr-fpmath:bf16 ,,768x768,0.138184`

jondea commented:
If you wanted a quick work around, you could pass in format any from PyTorch when you re-init the primitive even though you know it's going to be that format.

Nevertheless, this change does make sense in the long term. I think instead of hard coding it in oneDNN, it makes sense to check that the format passed into init matches the one returned from ACL. That would be more flexible to future changes in ACL.

snadampal commented:
Hi @jondea , given this is the right direction, I extended the supported formats to `any` and `blocked`. of course we can debate whether the blocked always guarantee that it's the one supported by ACL or not. for that I think we can rely one of the ACL config checks to return not supported error.

I didn't want to workaround ideep as it will be too hacky solution; since there won't be any on-the-fly reorders in the compiled case, we need to hack the weight md format at multiple places, during create as well as exec, which i don't think is a good option.

snadampal commented:
I'm trying to add few tests for ip in sbgemm mode. Could someone please point me to documentation on how to add benchdnn unit tests for new configurations? Thank you!

dzarukin commented on src/cpu/cpu_inner_product_list.cpp:
```diff
@@ -83,6 +83,10 @@ const std::map<pk_dt_impl_key_t, std::vector<impl_list_item_t>> &impl_list_map()
             CPU_INSTANCE(ref_inner_product_fwd_t)
             nullptr,
         }},
+        {{forward, f32, bf16, f32}, {
```
Could you, please, elaborate on why this configuration is needed?

Just in case, fpmath mode requires both src and weights to be fp32 to work. From functional point of view, the implementation takes care of down-converting and applying low precision computation.

snadampal replied:
Hi @dzarukin , thanks for the review.  your explanation of the fastmath mode is what we have been using so far for the eager mode execution of the model, where we create primitive with both src and weights in fp32, and then reorder weights (from fp32 to bf16) for every exec. 
However, with graph compilation, we are able to reorder/pre-pack all the weights during the model load and compilation phase itself so that redundant and on-the-fly reorders can be avoided. This PR is to support the compile scenario where src in fp32 and weights in bf16 for the primitive create. ACL primitives already support this mode (i call it sbgemm), but lacking this definition in this list is preventing us from creating them. 
I will add some of these comments to the source file.

jondea commented:
I think we need a bit more logic inside `init_conf_ip` Currently it is set up for just `format_tag::any`, and so with your changes it will accept `any` or any `blocked` format. I think in the case of `blocked` you also need to check that the format produced by `acl_utils::reorder_to_weight_format` matches the one passed in.

To be concrete, I think as it stands, this will fail benchdnn if you run with a `blocked` format that is not supported by ACL. Whereas it should return unimplemented and pass to ref.

snadampal commented:
Hi @jondea , addressed your feedback about the error checks for blocked format and also added unit tests. Please note that for supporting the unit test with `--wtag=Ab8a` blocked layout I had to set `--allow-enum-tags-only=0` and I hence I'm passing `allow_all_flags=true` to `set_default_params()`.  Please let me know if you see any issues with it. 
this is how I tested the blocked format
`./benchdnn --ip --mode=P --engine=cpu --allow-enum-tags-only=0 --batch=inputs/ip/test_ip_acl`

jondea commented on tests/benchdnn/inputs/ip/test_ip_acl:
```diff
@@ -24,3 +24,10 @@
 # 2d-spatial dimensions
 --batch=shapes_googlenet_v1
+
+# Tests for external blocked weights layout.
```
Thanks for adding these examples. Just to check, does this pass on Graviton 2, 3 (maybe even 4?)? The chosen (and therefore supported) blocked weight format depends on the micro architecture, so it might be worth putting this in a machine specific input file.

snadampal replied:
good point. but it will just skip on the unsupported platforms right. could you please elaborate what you mean by machine specific input file. do you mean adding a new input file something like `test_ip_acl_neoversev1` ?
Sorry, I'm extending the benchdnn unit tests for the first time, might be missing some basics.

snadampal replied:
currently the test is passing on Graviton3 and is skipped on Graviton2.

jondea replied:
Yes, I meant something like that, but I forgot how the skip mechanism works. That should work fine, thanks!

jondea commented on src/cpu/aarch64/acl_inner_product.hpp:
```diff
@@ -257,10 +268,17 @@ struct acl_inner_product_fwd_t : public primitive_t {
+            if ((weights_format_kind_received == format_kind::blocked)
+                    && !(dnnl_memory_desc_equal(
+                            &weights_md_received, &weights_md_))) {
+                return status::unimplemented;
```
Please could you use the `ACL_CHECK_SUPPORT` macro, along with some helpful words, maybe something like:
`specific blocked format not supported by ACL, use format_kind_t::any to find a supported blocked format for your platform`

jondea reviewed (changes requested):
Thanks for the changes! I just have two minor comments.

snadampal commented:
Looks like the CI failures are not related to this PR, I have seen CI failing for other PRs too,  with the same errors.
@vpirogov , Is this a known issue?

```
The following tests FAILED:
	180 - test_benchdnn_modeC_self_smoke_cpu (Failed)
Errors while running CTest
```

```
        Start 180: test_benchdnn_modeC_self_smoke_cpu

180: Test command: /home/vsts/work/1/s/build/tests/benchdnn/benchdnn "--mode=C" "-v1" "--engine=cpu" "--self" "--batch=test_self_smoke"
180: Working Directory: /home/vsts/work/1/s/build/tests/benchdnn
180: Test timeout computed to be: 10000000
180: check_simple_enums() ...
180: check_attr2str() ...
180: check_attr() ...
180: [int self::check_attr():109] '(entry.policy) == (policy_t::PER_OC)' FAILED ==>  0 != 1
180: check_post_ops2str() ...
180: check_str2post_ops() ...
180: check_tags() ...
180: check_trim_tags() ...
180: check_skip_impl() ...
180: check_status_change() ...
180: [   0][0] exp_f32:           0 exp:           0 got:          -1 diff:       1 rdiff:       1
180: [   1][1] exp_f32:           1 exp:           1 got:           0 diff:       1 rdiff:       1
180: [   2][2] exp_f32:           2 exp:           2 got:           1 diff:       1 rdiff:     0.5
180: [   3][3] exp_f32:           3 exp:           3 got:           2 diff:       1 rdiff:0.333333
180: [   4][4] exp_f32:           4 exp:           4 got:           3 diff:       1 rdiff:    0.25
180: [   5][5] exp_f32:           5 exp:           5 got:           4 diff:       1 rdiff:     0.2
180: [   6][6] exp_f32:           6 exp:           6 got:           5 diff:       1 rdiff:0.166667
180: [   7][7] exp_f32:           7 exp:           7 got:           6 diff:       1 rdiff:0.142857
180: [   8][8] exp_f32:           8 exp:           8 got:           7 diff:       1 rdiff:   0.125
180: [   9][9] exp_f32:           9 exp:           9 got:           8 diff:       1 rdiff:0.111111
180: [COMPARE_STATS]: trh=0 max_diff:       1 max_rdiff:       1
180: [   0][0] exp_f32:           0 exp:           0 got:          -1 diff:       1 rdiff:       1
180: [   1][1] exp_f32:           1 exp:           1 got:           0 diff:       1 rdiff:       1

180: check_norm() ...
180: check_diff_norm() ...
180: check_compare_norm() ...
180: [L0] = 1
180: [L1] exp:      55 got:      65 diff:      10 rel_diff:0.181818
180: [L2] exp: 19.6214 got: 22.4722 diff: 3.16228 rel_diff:0.161165
180: [L8] exp:      10 got:      11 diff:       1 rel_diff:     0.1
180: check_compare() ...
180: [  99][99] exp_f32:           1 exp:           1 got:         100 diff:      99 rdiff:      99
180: [COMPARE_STATS]: trh=98 max_diff:      99 max_rdiff:      99
180: tests:18 passed:17 skipped:0 mistrusted:0 unimplemented:0 invalid_arguments:0 failed:1 listed:0
180: total: 0.03s;
180/184 Test #180: test_benchdnn_modeC_self_smoke_cpu ......................***Failed    0.06 sec
```
```

igorsafo commented:
> Looks like the CI failures are not related to this PR, I have seen CI failing for other PRs too, with the same errors. @vpirogov , Is this a known issue?
> 
> ```
> The following tests FAILED:
> 	180 - test_benchdnn_modeC_self_smoke_cpu (Failed)
> Errors while running CTest
> ```
> 
> ```
>         Start 180: test_benchdnn_modeC_self_smoke_cpu
> 
> 180: Test command: /home/vsts/work/1/s/build/tests/benchdnn/benchdnn "--mode=C" "-v1" "--engine=cpu" "--self" "--batch=test_self_smoke"
> 180: Working Directory: /home/vsts/work/1/s/build/tests/benchdnn
> 180: Test timeout computed to be: 10000000
> 180: check_simple_enums() ...
> 180: check_attr2str() ...
> 180: check_attr() ...
> 180: [int self::check_attr():109] '(entry.policy) == (policy_t::PER_OC)' FAILED ==>  0 != 1
> 180: check_post_ops2str() ...
> 180: check_str2post_ops() ...
> 180: check_tags() ...
> 180: check_trim_tags() ...
> 180: check_skip_impl() ...
> 180: check_status_change() ...
> 180: [   0][0] exp_f32:           0 exp:           0 got:          -1 diff:       1 rdiff:       1
> 180: [   1][1] exp_f32:           1 exp:           1 got:           0 diff:       1 rdiff:       1
> 180: [   2][2] exp_f32:           2 exp:           2 got:           1 diff:       1 rdiff:     0.5
> 180: [   3][3] exp_f32:           3 exp:           3 got:           2 diff:       1 rdiff:0.333333
> 180: [   4][4] exp_f32:           4 exp:           4 got:           3 diff:       1 rdiff:    0.25
> 180: [   5][5] exp_f32:           5 exp:           5 got:           4 diff:       1 rdiff:     0.2
> 180: [   6][6] exp_f32:           6 exp:           6 got:           5 diff:       1 rdiff:0.166667
> 180: [   7][7] exp_f32:           7 exp:           7 got:           6 diff:       1 rdiff:0.142857
> 180: [   8][8] exp_f32:           8 exp:           8 got:           7 diff:       1 rdiff:   0.125
> 180: [   9][9] exp_f32:           9 exp:           9 got:           8 diff:       1 rdiff:0.111111
> 180: [COMPARE_STATS]: trh=0 max_diff:       1 max_rdiff:       1
> 180: [   0][0] exp_f32:           0 exp:           0 got:          -1 diff:       1 rdiff:       1
> 180: [   1][1] exp_f32:           1 exp:           1 got:           0 diff:       1 rdiff:       1
> 
> 180: check_norm() ...
> 180: check_diff_norm() ...
> 180: check_compare_norm() ...
> 180: [L0] = 1
> 180: [L1] exp:      55 got:      65 diff:      10 rel_diff:0.181818
> 180: [L2] exp: 19.6214 got: 22.4722 diff: 3.16228 rel_diff:0.161165
> 180: [L8] exp:      10 got:      11 diff:       1 rel_diff:     0.1
> 180: check_compare() ...
> 180: [  99][99] exp_f32:           1 exp:           1 got:         100 diff:      99 rdiff:      99
> 180: [COMPARE_STATS]: trh=98 max_diff:      99 max_rdiff:      99
> 180: tests:18 passed:17 skipped:0 mistrusted:0 unimplemented:0 invalid_arguments:0 failed:1 listed:0
> 180: total: 0.03s;
> 180/184 Test #180: test_benchdnn_modeC_self_smoke_cpu ......................***Failed    0.06 sec
> ```

Hi @snadampal  Thanks for notifying, we are aware of this issue and a fix is submitted into the internal repository.

snadampal commented:
@igorsafo , thanks for confirming. I guess CI will be triggered for this PR once the internal issue is fixed, right?
This PR has been approved and I think it's ready for merge post CI runs. Please let me know if there are any additional steps. thank you!

igorsafo commented:
> @igorsafo , thanks for confirming. I guess CI will be triggered for this PR once the internal issue is fixed, right? This PR has been approved and I think it's ready for merge post CI runs. Please let me know if there are any additional steps. thank you!

Once the fix is landed we will need to rebase this branch on top of the latest `main` to include the fix, otherwise CI will stay red. I will notify once the fix is available.

igorsafo commented:
@snadampal The fixes landed into the `main`. Could you please rebase the branch?

snadampal commented:
Hi @igorsafo , done, rebased the PR.

snadampal commented:
Hi @igorsafo , please let me know if anything else to be covered for this PR to get merged. thank you.

snadampal commented:
Hi @igorsafo , we planned this PR for oneDNN 3.4. It was tagged for v3.4 but not yet merged,  it's been approved and CI passing, could you please prioritize it?

snadampal commented:
thank you!

igorsafo commented:
Hi @snadampal , sorry for the wait, the changes were promoted into main and backported into rls-v3.4.
Thank you for the contribution!
